### DISCUSSION TITLE
How to manage fonts in web assembly?

### DISCUSSION THREAD

**Discussion by appel1** [AvaloniaUI/Avalonia#15131] [Q&A] [answered]
Until #14941 is completed and browsers support it, you have to load all fonts you need yourself. In the simple cases when you know beforehand what you need and only need a few ttf files you can embed them and load on startup, but as soon as you want to support arbitrary text input you end up with quite a lot of data. If you then need variants like italic, bold and so on it gets even worse. For just Noto Sans SC you end up with 90MB. Variable fonts, #11092, could help reduce the size but it still quite a lot. Ideally you want to support letting the user pick fonts for certain things which makes embedding all font data even less desirable. 

One possible solution is to load fonts on demand in a IFontCollection implementation, but then you run into the issue that font manager isn't asynchronous and you can't download anything synchronously afaik in web assembly. Is there a way to do that that I've missed?

Has anyone else solved this in some way?

Looking at where fonts are used in Avalonia it doesn't seem feasible to make it fully async. 

My current idea is to load fonts on demand and then when they've been loaded invalidate the whole visual tree to ensure all text is re-rendered but with the new font data.

**Comment by dbriard** [comment, marked as answer]
Personally, I load fonts on demand from Google Fonts and render using a fallback font during the downloading process. Then, I invalidate the rendering to display the correct font.

For the first (default) font, I can delay the rendering until the font is loaded.

I also have a font combo box so that the user can change the font in the app. When the user selects a font in the combo box, I do not immediately raise the font change. Instead, I download the font and then raise the font change when I know that the font is available.

Ideally, the fallback fonts are almost never displayed as fonts are part of user documents and they are loaded async (including the fonts).

I do not know about Avalonia's FontManager, as I am using a custom made text control based on RichTextKit.

**Reply by Gillibald** [reply]
This works with Avalonia as well. It is possible to populate a font collection via streams. So it should be possible to download a web font via HttpClient at startup etc. The downloaded font should be cached by the browser so you only need to do this once.

**Comment by appel1** [comment]
Yes, looks like it works fairly well to load the list of fonts at startup and then download fonts on demand followed by invalidating the whole visual tree when the download is completed. Still have to figure out how to make popups invalidate and the scroll implementation doesn't seem to like it when content changes size while scrolling. 

An outstanding issue I've yet to figure out is to provide different fallback fonts depending on the requested font. For example if the user picks Noto Serif and the rendered text contains Japanese but my list of fallback fonts is Noto Sans *. Possibly something for a feature request. 

```csharp
public static class AppBuilderExtension
{
    public static AppBuilder WithAbacusFonts(this AppBuilder appBuilder, System.Net.Http.HttpClient client, WebFontList? fontList)
    {
        Uri key = new Uri("fonts:SystemFonts");
        appBuilder.ConfigureFonts(fontManager
            => fontManager.AddFontCollection(new FontsCollection(client, key, fontList)));
        appBuilder.With(new FontManagerOptions
        {
            DefaultFamilyName = "Noto Sans",
            FontFallbacks = fontList?.Items
                .Where(item => item.Family.StartsWith("Noto Sans ", StringComparison.Ordinal))
                // TODO: Create UnicodeRanges based on subsets?
                .Select(item => new FontFallback { FontFamily = item.Family })
                .ToList(),
        });

        appBuilder.AfterSetup(ab =>
        {
            if (FontManager.Current.SystemFonts is not FontsCollection f)
                return;

            f.TypefaceLoaded += (s, a) =>
            {
                switch (Application.Current?.ApplicationLifetime)
                {
                    case IClassicDesktopStyleApplicationLifetime desktop:
                        foreach (var w in desktop.Windows)
                        {
                            LayoutHelper.InvalidateSelfAndChildrenMeasure(w);
                        }

                        break;
                    case ISingleViewApplicationLifetime { MainView: not null } single:
                        LayoutHelper.InvalidateSelfAndChildrenMeasure(single.MainView);
                        break;
                }
            };
        });

        return appBuilder;
    }
}
```
